api/src/main/java/org/apache/iceberg/actions/DeleteOrphanFiles.java

```diff
@@ -67,7 +67,11 @@ public interface DeleteOrphanFiles extends Action<DeleteOrphanFiles, DeleteOrpha
    *
    * @param deleteFunc a function that will be called to delete files
    * @return this for method chaining
+   * @deprecated Deletes are now performed in bulk see {@link #deleteBulkWith(Consumer)}. This
```

Comment:
This will break the public API if a table is configured with a non-bulk IO and the user passes a custom delete function. In a lot of cases, a custom delete simply adds elements to a set. After this change, the custom delete function may be ignored and we will actually delete files. 

Do we actually need to attempt bulk deletes in if a custom function is provided? I'd consider just going through the non-bulk API in that case. That can simplify the implementation.

```
private Consumer<String> deleteFunc = null;

...

@Override
public DeleteOrphanFilesSparkAction deleteWith(Consumer<String> newDeleteFunc) {
  this.deleteFunc = newDeleteFunc;
  return this;
}

...

private void deleteFiles(Iterable<String> paths) {
  if (deleteFunc == null && table.io() instanceof SupportsBulkOperations) {
    SupportsBulkOperations bulkIO = (SupportsBulkOperations) table.io();
    if (deleteExecutorService != null) {
      bulkIO.deleteFiles(deleteExecutorService, paths);
    } else {
      bulkIO.deleteFiles(paths);
    }

  } else {
    Tasks.foreach(paths)
        .noRetry()
        .executeWith(deleteExecutorService)
        .suppressFailureWhenFinished()
        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
        .run(deleteFunc != null ? deleteFunc::accept : table.io()::deleteFile);
  }
}
```

Reply:
I believe the above snippet would be backward compatible and would support the existing API?

Reply:
One of my goals here is to remove those existing apis. I feel like custom delete function here is solely here for testing. Any user who actually wants to deal with their results would use the "toDataframe" option for Spark.

Currently we have too many options here and no real use cases. This really shouldn't be a set of configuration or options that users should be concerned with. If they are, the correct spot to be modifying this is the FileIO. For example, if I have multiple actions running at the same time, the coordination of requests to S3 or whatnot should happen at the FileIO, otherwise the user needs to manually be determining the correct values for each of their possible concurrent jobs.

Reply:
An empty delete function is also used for dry runs to analyze candidates prior to their physical removal. We can also collect go the data frame route but we offer `deleteWith` in a number of APIs, so I don't see a big problem.

> For example, if I have multiple actions running at the same time, the coordination of requests to S3 or whatnot should happen at the FileIO, otherwise the user needs to manually be determining the correct values for each of their possible concurrent jobs.

We should have a way to configure this global default in `FileIO` and it is a correct place to do it. My point is that I don't see harm allowing certain operations to override that default, especially if it makes the implementation simpler.

Reply:
My primary worry is the complexity in the code caused by the deprecation.